File: 45. Jump Game II.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        '''
        the thought is:
        1. get the nearest and the furthest pos from the last step
        2. calculate the furthest of this step
        3. the nearest of this step should be the futrhest from last + 1
        '''
        if len(nums) <= 1:
            return 0
        start, end, maxend = 0, 0, 0
        step = 0
        while True:
            step += 1
            for i in range(start, end+1):
                if i + nums[i] >= len(nums)-1:
                    return step
                maxend = max(maxend, i+nums[i])
            start, end = end+1, maxend

File: 45. Jump Game II.java
/**
 * keep track of this step max and next step max,
 * we try to find the max of each step,
 * and we check one by one in each step
 * Time O(n)
 * Space O(1)
 */
class Solution {
    public int jump(int[] nums) {
        if (nums==null || nums.length==0) {
            return 0;
        }
        int n = nums.length;
        int thismax=0, nextmax=0, step=0;
        for (int i=0; i<n-1; i++) {
            nextmax = Math.max(nextmax, i+nums[i]);
            if (i == thismax) {
                thismax = nextmax;
                step++;
            }
        }
        return step;
    }
}